updater.go

```diff
@@ -126,23 +130,31 @@ func (u *Updater) update(ctx Context, options UpdateOptions) (*Update, *Error) {
 		return update, verifyErrPtr(err)
 	}
 
-	if err := ctx.BeforeApply(*update); err != nil {
-		return update, applyErrPtr(err)
+	if err := u.apply(ctx, *update, options); err != nil {
+		return update, err
 	}
 
-	if err := u.platformApplyUpdate(*update, options); err != nil {
-		return update, applyErrPtr(err)
+	if err := ctx.Restart(); err != nil {
+		return update, restartErrPtr(err)
```

Comment:
Yeah I suppose the golang way is to use the error interface everywhere and to use a type switch when you need to check for specific error type?

It always seemed "code smell" to be checking types and casting but I suppose in go since you don't have inheritance using type switches isn't a bad thing?